src/lib.rs

```diff
@@ -1,7 +1,31 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
-    }
+pub mod dummy;
+pub mod openssl;
+
+/// A trait for a Verifiable Random Functions (VRF) implementations.
+pub trait VRF<PublicKey, SecretKey> {
+    type Error;
+
+    /// Generates proof from a secret key and message
+    ///
+    /// # Arguments
+    ///
+    /// * `x`     - A secret key
+    /// * `alpha` - A slice representing the message in octets
+    ///
+    /// # Returns
+    ///
+    /// * If successful, a vector of octets representing the proof of the VRF
+    fn prove(&mut self, x: SecretKey, alpha: &[u8]) -> Result<Vec<u8>, Self::Error>;
```

Comment:
```suggestion
    fn prove(&self, x: SecretKey, alpha: &[u8]) -> Result<Vec<u8>, Self::Error>;
```

I think it would make more sense to have `prove` and `verify` as immutable methods, but from what I see in the code the `&mut` is needed because of the `BigNumContext`...

Reply:
Indeed.

Reply:
The `openssl` crate API is... not very ergonomic.

Reply:
totally agree... :disappointed: